### DISCUSSION TITLE
MME rejected gNB registration

### DISCUSSION THREAD

**Discussion by bocob** [magma/magma#15260] [Q&A]
Hi, 

I am trying to integrate gNB with magma5g core, but I got the following in mme.log indicated AMF setup failure. the gNB ID, TAC were matched with the org8r provisioned. gNB is unmanaged. any insight what might be the problem?

No Common PLMN with gNB, generate_ng_setup_failure : -1

thanks.

**Comment by panyogesh** [comment]
@bocob : Can you please share the pcap of generated packet & output of /var/opt/magma/tmp/mme.conf

**Reply by bocob** [reply]
@panyogesh thanks for your reply. here are the logs.
[logs.zip](https://github.com/magma/magma/files/12068475/logs.zip)

**Comment by panyogesh** [comment]
[15260-Comparision-Working-NonWorking.xlsx](https://github.com/magma/magma/files/12072759/15260-Comparision-Working-NonWorking.xlsx)

**Comment by panyogesh** [comment]
Hi @bocob : Pcap shared looks interesting. It does not have a PLMN (MCC/MNC) during the setup Request. I am attaching an excel comparing Non-Working& Working pcap back to back.

We have tested 5G with simulators and real equipment and somehow all of them sends the PLMN in Initial Setup.

Can you please check and let us know little bit more about the tool you are using.

**Reply by bocob** [reply]
Hi @panyogesh , there was PLMN in the setup request message. please see the attached pic for your reference.

![image](https://github.com/magma/magma/assets/40008962/a118de02-f8ab-4870-80ee-5630e1cb0db0)

**Comment by panyogesh** [comment]
@bocob : The issue seems to be in Slice Differentiator configuration 
![image](https://github.com/magma/magma/assets/69527565/629b6639-1655-4ebe-901c-db50f944a754)

{ MCC="001" ; MNC="01"; AMF_DEFAULT_SLICE_SERVICE_TYPE="1" ; AMF_DEFAULT_SLICE_DIFFERENTIATOR="ffffff"; }

If we make Slice Diffrentiator same (ffffff -> 000000) it should go through. Can you plese check and confirm.

**Reply by bocob** [reply]
Hi @panyogesh, thanks! I got much farther now. UE attached, but 5G core was not assigning IP due to " DNN is not Supported or not Subscribed, reject with a cause: 27". from wireshark log, "DNN: internet" was sent to 5gcore. UE subscriber has "internet". what else is missing? thanks.
![image](https://github.com/magma/magma/assets/40008962/ae9924c4-4c03-440d-b2b2-7c05991e484c)

**Reply by Radula76** [reply]
Ue might be requesting ipv4v6 APN?

**Reply by bocob** [reply]
the UE was requesting ipv4.

**Comment by panyogesh** [comment]
@bocob I have tested the configuration on new builds. Seems to be ok. Can you please share the pcap and mme.log
![Uploading image.png…]()

**Comment by panyogesh** [comment]
<img width="960" alt="internet-dnn-Snapshot" src="https://github.com/magma/magma/assets/69527565/ac8df29e-6912-4b51-89ef-a0fbe509b17b">

**Reply by bocob** [reply]
hi @panyogesh, after restarting the AGW, UE was able to attach and got IP. I see another issue that after inactivity timer expired, UE would not re-establish the session. 5gcore keeps release the session. have you seen this before?
![image](https://github.com/magma/magma/assets/40008962/b1713e87-7ea7-4788-9276-f23a82bb5c49)

**Reply by bocob** [reply]
[logs-2023-07-19.zip](https://github.com/magma/magma/files/12099985/logs-2023-07-19.zip)

**Comment by panyogesh** [comment]
@bocob : I will check and update.

**Comment by panyogesh** [comment]
@bocob : By chance do you have the syslog for this failure. I am suspecting some pending cleanup.

**Reply by bocob** [reply]
@panyogesh here is the syslog...
[syslog.tar.gz](https://github.com/magma/magma/files/12143012/syslog.tar.gz)

**Reply by bocob** [reply]
@panyogesh after rebooting the host, the core is still acting the same way; can not re-establish a session. do you think if it was a configuration issue or a debug in the core?  thanks.

**Reply by panyogesh** [reply]
Can you please share a mme.log. Also are you using some open source simulator. I just want a way to reproduce in my setup.

**Reply by bocob** [reply]
@panyogesh here are the mme.log. the gNB is a real gNB. the gNB was working fine with other OEM 5g core.
[mme.log.tar.gz](https://github.com/magma/magma/files/12187570/mme.log.tar.gz)

**Reply by Radula76** [reply]
There are a lot of cleanup issues on magma currently.  when the Ue goes idle we need to make sure the proper 5G states are maintained.  it is very different to LTE.  some radio's allow you to extend the Ue idle timer which is helpful during testing.  The core can receive a new session request and then tries to remove the old session information but it is in the wrong state, it will get into a loop.  we can use the command "sudo config_stateless_agw.py flushall_redis" to clear redis and restart the core application, the Ue connection request will be rejected and a fresh session setup will work (as well as sctp reset).

**Reply by bocob** [reply]
@Radula76 thanks for your info. is that mean currently magma would not support session re-establishment? when would this fix(cleanup)? thanks.

**Reply by Radula76** [reply]
are you using magma 1.8 or the latest mater build (1.9)?   1.9 should be working in terms of idle to active session re-establishment.
the 5G states are different to LTE: http://howltestuffworks.blogspot.com/2019/09/5g-nr-ue-rrc-states-and-state.html
the 2nd half of the page explains the mobility management.
active to idle and back again should be working.  if the gNB powers off its a bit harder for the AGW to clear out the active sessions.  the idle sessions can be cleared out, but there are a few longer timers on the AGW to try and hold onto the active Ue session.  if the Ue comes back with old session id's the core can get confused.

**Reply by bocob** [reply]
thanks @Radula76  

I was using 1.8

**Comment by panyogesh** [comment]
1. Regarding the Slice-Differentiator Value set to 0 and NgSETUp not working as expected.
<**Response**> As per the specs the default value of Slice Differentiator is 0xffff. If testing is done with SD=0, it needs to be set in
           AMF. It can be done via NMS or Swagger snapshots attached in Zip. Refer SD-0.zip. 

2. Issue with Idle mode translation
<**Response**> Logs for the Testcase similar to working idle mode is attached (Idle-Timeout-Logs)

3. The issue reported : I see another issue that after inactivity timer expired, UE would not re-establish the session. 5gcore keeps release the session. have you seen this before?
<**Response**> The packet 35444 highlighted seems to be normal release. I guess once the Slice-Differentiator being 0 issue is resolved please share the pcaps, mme.log and syslog for checking the issue. The older logs send seems little out of sync. 
[Idle-Timeout-Logs.zip](https://github.com/magma/magma/files/12207866/Idle-Timeout-Logs.zip)
[SD-0.zip](https://github.com/magma/magma/files/12207867/SD-0.zip)

**Reply by bocob** [reply]
@panyogesh since July 19 post, UE was able to do initial attach. SD was updated to "0"(otherwise UE was not able to do initial attach). as @Radula76 explained, ver 1.8 does not support RRC-idle to RRC-connect. any idea when is going to support gNB reset and 5gcore cleanup properly?  thanks.

**Comment by panyogesh** [comment]
@bocob: Idle mode support is already part of 1.8 release. The tested logs with real equipment is attached in 1.8 release link[release link](https://github.com/magma/magma/releases/tag/v1.8.0) . 
Under 5G-SA:[ Test-Logs](https://drive.google.com/drive/folders/1Edk01f0tTAtH5r8qK-t1phxuvOv9GPqd). 
Inside that in the real equipment section [logs](https://drive.google.com/drive/folders/1Wy1iwWOGnqewDFBiHq2JXsipVOEzym6x).

When compared to the pcaps you shared there seems to be 2 issues:
  1. GNB is not sending context response (might be UE is not confirming)
  2. Second service request is not accepted by AGW,
 
For issue 1: If you can compare the working & non-working pcap of packet 35601 on your setup and share the details it will be great.
                   Some UEs are slightly touchy about few IE.
For issue 2: It will be great if you can share the mme.log, syslog and pcap when the issue occurred.

For both I am attaching the comparison snapshot.
[Uploading IDLE-MODE-FEATURE.zip…]()

**Reply by panyogesh** [reply]
@bocob: gnB reset is also part of 1.8 but may be there are specific scenarios that needs to be verified. If you have issue logs related to that I can check it.

**Reply by bocob** [reply]
@panyogesh: I have let the 5gcore idle for few days and revisited it today, I found sctpd did not respond to INIT. so I tried to clean up by running "sudo config_stateless_agw.py flushall_redis". after reboot the whole box, now MME is not even running... any idea why and how to recover?

root@magma:/var/log# sudo service magma@mme  status
● magma@mme.service - Magma OAI MME service
     Loaded: loaded (/etc/systemd/system/magma@mme.service; disabled; vendor preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Thu 2023-08-03 19:56:28 UTC; 4s ago
    Process: 128874 ExecStartPre=/usr/bin/env python3 /usr/local/bin/generate_oai_config.py (code=exited, status=0/SUCCESS)
    Process: 128946 ExecStartPre=/usr/bin/env python3 /usr/local/bin/config_stateless_agw.py reset_sctpd_for_stateful (code=exited, status=0/SUCCESS)
    Process: 128952 ExecStart=/usr/local/bin/mme -c /var/opt/magma/tmp/mme.conf -s /var/opt/magma/tmp/spgw.conf (code=exited, status=23)
    Process: 128953 ExecStopPost=/usr/bin/env python3 /usr/local/bin/service_util.py mme (code=exited, status=0/SUCCESS)
   Main PID: 128952 (code=exited, status=23)
root@magma:/var/log#


root@magma:/var/log# journalctl -n40 -u magma@mme
-- Logs begin at Thu 2023-07-13 10:14:47 UTC, end at Thu 2023-08-03 19:59:48 UTC. --
Aug 03 19:59:32 magma mme[143696]: WARNING:root:Service (spgw) missing in mconfig
Aug 03 19:59:32 magma mme[143709]: STATELESS        mme -> use_stateless
Aug 03 19:59:32 magma mme[143709]: STATELESS        pipelined -> clean_restart
Aug 03 19:59:32 magma mme[143709]: STATELESS        pipelined -> redis_enabled
Aug 03 19:59:32 magma mme[143709]: STATELESS        sessiond -> support_stateless
Aug 03 19:59:32 magma mme[143709]: Check returning return_codes.STATELESS
Aug 03 19:59:32 magma mme[143709]: AGW is stateless, no need to restart Sctpd
Aug 03 19:59:32 magma systemd[1]: Started Magma OAI MME service.
Aug 03 19:59:32 magma mme[143713]: ==143713==Sanitizer CHECK failed: ../../../../src/libsanitizer/lsan/lsan_interceptors.cpp:53 ((!lsan_init_is_running)) != (0) (0, 0)
Aug 03 19:59:32 magma systemd[1]: magma@mme.service: Main process exited, code=exited, status=23/n/a
Aug 03 19:59:33 magma systemd[1]: magma@mme.service: Failed with result 'exit-code'.
Aug 03 19:59:38 magma systemd[1]: magma@mme.service: Scheduled restart job, restart counter is at 164.
Aug 03 19:59:38 magma systemd[1]: Stopped Magma OAI MME service.
Aug 03 19:59:41 magma systemd[1]: Starting Magma OAI MME service...
Aug 03 19:59:41 magma mme[144323]: ERROR:root:Error retrieving config for mme, key not found: enable_nat
Aug 03 19:59:41 magma mme[144323]: ERROR:root:Error retrieving config for mme, key not found: congestion_control_enabled
Aug 03 19:59:41 magma mme[144323]: ERROR:root:Error retrieving config for mme, key not found: enable5g_features
Aug 03 19:59:41 magma mme[144323]: ERROR:root:Error retrieving config for mme, key not found: amf_default_slice_service_type
Aug 03 19:59:41 magma mme[144323]: ERROR:root:Error retrieving config for mme, key not found: amf_default_slice_differentiator
Aug 03 19:59:41 magma mme[144323]: ERROR:root:Error retrieving config for mme, key not found: amf_name
Aug 03 19:59:41 magma mme[144323]: ERROR:root:Error retrieving config for mme, key not found: amf_region_id
Aug 03 19:59:41 magma mme[144323]: ERROR:root:Error retrieving config for mme, key not found: amf_set_id
Aug 03 19:59:41 magma mme[144323]: ERROR:root:Error retrieving config for mme, key not found: amf_pointer
Aug 03 19:59:41 magma mme[144323]: ERROR:root:Error retrieving config for mme, key not found: default_dnn
Aug 03 19:59:41 magma mme[144323]: ERROR:root:Error retrieving config for mme, key not found: auth_retry_max_count
Aug 03 19:59:41 magma mme[144323]: ERROR:root:Error retrieving config for mme, key not found: auth_retry_interval
Aug 03 19:59:41 magma mme[144323]: ERROR:root:Error retrieving config for spgw, key not found: ovs_internal_sampling_fwd_tbl
Aug 03 19:59:41 magma mme[144323]: WARNING:root:Service (spgw) missing in mconfig
Aug 03 19:59:41 magma mme[144337]: STATELESS        mme -> use_stateless
Aug 03 19:59:41 magma mme[144337]: STATELESS        pipelined -> clean_restart
Aug 03 19:59:41 magma mme[144337]: STATELESS        pipelined -> redis_enabled
Aug 03 19:59:41 magma mme[144337]: STATELESS        sessiond -> support_stateless
Aug 03 19:59:41 magma mme[144337]: Check returning return_codes.STATELESS
Aug 03 19:59:41 magma mme[144337]: AGW is stateless, no need to restart Sctpd
Aug 03 19:59:41 magma systemd[1]: Started Magma OAI MME service.
Aug 03 19:59:41 magma mme[144339]: ==144339==Sanitizer CHECK failed: ../../../../src/libsanitizer/lsan/lsan_interceptors.cpp:53 ((!lsan_init_is_running)) != (0) (0, 0)
Aug 03 19:59:41 magma systemd[1]: magma@mme.service: Main process exited, code=exited, status=23/n/a
Aug 03 19:59:41 magma systemd[1]: magma@mme.service: Failed with result 'exit-code'.
Aug 03 19:59:46 magma systemd[1]: magma@mme.service: Scheduled restart job, restart counter is at 165.
Aug 03 19:59:46 magma systemd[1]: Stopped Magma OAI MME service.
root@magma:/var/log#

**Reply by panyogesh** [reply]
This issue is coming due to a new Memory Leak library that got released. I am working with CI/CD team to get a build integrated with new library. Will share you the details as soon as its ready.

**Reply by bocob** [reply]
@panyogesh but I rebooted the whole box, would the memory leak be cleared? it does not seem like memory leak to me with "ERROR:root:Error retrieving config for mme, key not found: enable_nat"... it seems database without key to me... would it be caused by "sudo config_stateless_agw.py flushall_redis"?

**Reply by panyogesh** [reply]
@bocob : It was little tricky. Even I started with the same analysis. I reproduced same issue in my setup. After the upgrade of library the issue was resolved.

**Reply by bocob** [reply]
@panyogesh thx. how do I upgrade the library and recover from this situation?

**Reply by bocob** [reply]
@panyogesh would it be easier to reinstall AGW to recover from my current situation? or doing library upgrade as you did?

**Reply by panyogesh** [reply]
Unfortunately the package is broken. So it will be an upgrade to a broken package. We have decided to spin up the new images but it may take a weeks time. If you want I can share the private package or temporary steps to recover.

**Reply by bocob** [reply]
@panyogesh yes, please share the private package and temporary steps to recover. thanks.

**Reply by panyogesh** [reply]
wget https://ftp.debian.org/debian/pool/main/g/gcc-10/liblsan0_10.2.1-6_amd64.deb
wget https://ftp.debian.org/debian/pool/main/g/gcc-10/gcc-10-base_10.2.1-6_amd64.deb
sudo dpkg -i gcc-10-base_10.2.1-6_amd64.deb liblsan0_10.2.1-6_amd64.deb

Package I will share tomorrow

**Reply by bocob** [reply]
@panyogesh thanks. what else need to be done beside installing the 2 packages?

**Reply by bocob** [reply]
@panyogesh do I need to install "gcc-10-base_10.2.1-6_amd64.deb"? it only contains changelog file and a README file.

**Reply by bocob** [reply]
@panyogesh I have applied the lib packages and it is back to where I was. now I am still facing 5gcore failed with RRC-inactive to RRC-active. I have attached the log for a different vendor 5gcore and it was working. when comparing the "Service accept" message, the 5gcore sent few items. packet #29255 in working.pcap vs. packet #35601 in magma-17.pcap.
[logs-2023-08-10.zip](https://github.com/magma/magma/files/12317331/logs-2023-08-10.zip)
![image](https://github.com/magma/magma/assets/40008962/1991b239-1a78-4120-b224-b6132e33dd40)

**Reply by panyogesh** [reply]
Sure. I will check and update.

**Reply by panyogesh** [reply]
Found the diff between working and non-working. Is it possible to test on private binary

**Reply by bocob** [reply]
@panyogesh sure. I can test a private binary.

**Reply by shashidhar-patil** [reply]
Hi @bocob, 
From pcap comparison, we found one IE diff between working and non-working. 
We have raised a draft PR with the fix (https://github.com/magma/magma/pull/15312)
It would be great if you could test this fix

We have tested it with UERANSIM simulator.
Attached logs and pcap: [15260.zip](https://github.com/magma/magma/files/12827612/15260.zip)

**Reply by bocob** [reply]
hi @shashidhar-patil,
the problem that I was facing was not registration; initial registration was fine, but after UE went to RRC-inactive, UE failed to re-establish(resume) the session again, so RRC-inactive to RRC-active was failing.

**Reply by shashidhar-patil** [reply]
Hi @bocob,
We have reproduced the sevice request scenario. We have attached a complete pcap from registration flow to successful service request.
At packet number: 59 Ue moves to inactive state. 
At packet number: 63 Ue sends a service request, and core successfully responds to this request. Here Ue moves from rrc-inactive to rrc-active state.

**Reply by bocob** [reply]
hi @shashidhar-patil, is there an upgrade that I can do to resolve my problem from rrc-inactive to rrc-active?

**Reply by panyogesh** [reply]
@bocob : We will update you shortly.

**Reply by mickymkumar** [reply]
Hi @bocob, 

Please let us know if you still need upgrade steps for rrc-inactive to rrc-active or any other Support?
 
Thank you and regards,
Micky Kumar
